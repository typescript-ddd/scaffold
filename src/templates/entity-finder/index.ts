export * from "./entity-finder.generator";