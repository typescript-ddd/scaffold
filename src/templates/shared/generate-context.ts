import { SourceFile } from "ts-morph";

export type ContextImportDeclaration = { moduleSpecifier: string; namedImports: string[] };

export interface GenerateContext {
    resolveDir(type: "domain" | "infrastructure" | "application" | "presentation", subPath?: string): string;
    fileName: string;
    addImportDeclaration(declaration: ContextImportDeclaration): void;
    getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[];
    getModuleSpecifiers(): string[];
    getNamedImports(moduleSpecifier?: string): string[];    
    writeSourceFileImports: (sourceFile: SourceFile) => void;
    clearImportDeclarations(): void;
    tokenize: (text: string, tokens: { name: string }[]) => string;
}