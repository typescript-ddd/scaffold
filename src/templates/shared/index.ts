export * from "./context-builder";
export * from "./generate-context";