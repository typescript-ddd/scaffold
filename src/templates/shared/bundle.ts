import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
