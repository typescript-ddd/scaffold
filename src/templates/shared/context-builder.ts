import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
