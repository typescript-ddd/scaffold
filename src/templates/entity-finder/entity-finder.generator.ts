import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { EntityFinderTemplateValues } from "./entity-finder.types";

export const generateEntityFinder = (
  values: EntityFinderTemplateValues,
  context: GenerateContext
) => {
  const { entityName } = values;
  const entityFinderType = `${strings.capitalize(entityName)}Finder`;
  const repositoryType = `${strings.capitalize(entityName)}Repository`;
  const entityType = strings.capitalize(entityName);
  const entityIdType = `${entityType}Id`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${entityFinderType} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Criteria", "DomainEventBus", "EntityNotFoundError", "Uuid"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType, entityIdType, repositoryType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(entityFinderType);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(entityFinderType)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [
      {
        name: "repository",
        type: repositoryType,
        scope: Scope.Private,
        isReadonly: true,
      },
      {
        name: "domainEventBus",
        type: "DomainEventBus",
        scope: Scope.Private,
        isReadonly: true,
      },
    ],
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      entityFinderType
    )}.`,
    tags: [
      {
        tagName: "param",
        text: `{${repositoryType}} repository - The ${strings.natural(
          entityName
        )} repository.`,
      },
      {
        tagName: "param",
        text: `{DomainEventBus} domainEventBus - The domain event bus.`,
      },
    ],
  });

  classDeclaration
    .addMethod({
      name: "findById",
      scope: Scope.Public,
      returnType: `Promise<${entityType}>`,
      parameters: [
        {
          name: "id",
          type: entityIdType,
        },
      ],
      isAsync: true,
    })
    .setBodyText((writer) => {
      writer.writeLine(
        `const ${strings.lower(
          entityName
        )} = await this.repository.getById(id);`
      );
      writer.writeLine(`if (!${strings.lower(entityName)}) {`);
      writer.writeLine(`  throw new EntityNotFoundError("${entityType}", id);`);
      writer.writeLine(`}`);
      writer.writeLine(`return ${strings.lower(entityName)};`);
    })
    .addJsDoc({
      description: `Finds ${strings.describe(entityName)} by ${strings.describe(
        entityName
      )} identifier.`,
      tags: [
        {
          tagName: "param",
          text: `{${entityIdType}} id - The identifier of the ${strings.natural(
            entityName
          )}.`,
        },
        {
          tagName: "returns",
          text: `{Promise<${entityType}>} The ${strings.natural(entityName)}.`,
        },
      ],
    });

  classDeclaration
    .addMethod({
      name: "findAll",
      parameters: [{ name: "criteria", type: "Criteria" }],
      isAsync: true,
      returnType: `Promise<${entityType}[]>`,
    })
    .setBodyText((writer) => {
      const instances = strings.plural(strings.lower(entityName));
      writer.writeLine(
        `const ${instances} = await this.repository.getAll(criteria);`
      );
      writer.writeLine(`return ${instances};`);
    })
    .addJsDoc({
      description: `Finds all ${strings.plural(
        strings.natural(entityName)
      )} with the specified criteria.`,
      tags: [
        {
          tagName: "param",
          text: `{Criteria} criteria - The criteria to find the ${strings.plural(
            strings.natural(entityName)
          )}.`,
        },
        {
          tagName: "returns",
          text: `{Promise<${entityType}[]>} The ${strings.plural(strings.natural(entityName))} matching the criteria.`,
        },
      ],
    });

  let output = sourceFile.getFullText();

  return output;
};
