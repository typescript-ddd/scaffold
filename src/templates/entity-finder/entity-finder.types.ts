import { TemplateStringOption } from "../shared/template";

export type EntityFinderTemplateValues = {
  entityName: string;
};

export type EntityFinderTemplateOptions = {
  entityName: TemplateStringOption;
};
