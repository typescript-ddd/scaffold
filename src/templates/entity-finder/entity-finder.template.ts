import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateEntityFinder } from "./entity-finder.generator";
import {
  EntityFinderTemplateOptions,
  EntityFinderTemplateValues,
} from "./entity-finder.types";

export class EntityFinderTemplate
  implements Template<EntityFinderTemplateValues, EntityFinderTemplateOptions>
{
  readonly name = "Entity Finder";
  readonly description = "Generates an entity finder";
  readonly options: EntityFinderTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
  };
  readonly defaultValues = {
    entityName: "User",
  };
  generate(
    values: EntityFinderTemplateValues,
    context: GenerateContext,
    chunkName?: string
  ): Chunk {
    return generateEntityFinder(values, context, chunkName);
  }
}
