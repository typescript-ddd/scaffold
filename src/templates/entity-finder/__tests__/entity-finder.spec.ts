import { GenerateContext, ContextBuilder } from "../../shared";
import { generateEntityFinder } from "../entity-finder";

describe("EntityFinder", () => {
  let context: GenerateContext;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
  });

  it("Should render the entity finder template", () => {
    const output = generateEntityFinder(
      {
        entityName: "User",
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
