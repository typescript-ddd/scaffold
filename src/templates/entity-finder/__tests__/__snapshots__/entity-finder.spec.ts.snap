// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityFinder Should render the entity finder template 1`] = `
"import { Criteria, DomainEventBus, EntityNotFoundError, Uuid } from "@typescript-ddd/core";
import { User, UserId, UserRepository } from "@src/domain";

/** Represents a user finder. */
export class UserFinder {
    /**
     * Initializes a new instance of a user finder.
     * @param {UserRepository} repository - The user repository.
     * @param {DomainEventBus} domainEventBus - The domain event bus.
     */
    private constructor(private readonly repository: UserRepository, private readonly domainEventBus: DomainEventBus) {
    }

    /**
     * Finds a user by a user identifier.
     * @param {UserId} id - The identifier of the user.
     * @returns {Promise<User>} The user.
     */
    public async findById(id: UserId): Promise<User> {
        const user = await this.repository.getById(id);
        if (!user) {
          throw new EntityNotFoundError("User", id);
        }
        return user;
    }

    /**
     * Finds all users with the specified criteria.
     * @param {Criteria} criteria - The criteria to find the users.
     * @returns {Promise<User[]>} The users matching the criteria.
     */
    async findAll(criteria: Criteria): Promise<User[]> {
        const users = await this.repository.getAll(criteria);
        return users;
    }
}    
"
`;
